File: src/cpp/MoreThanHalfNum.cpp
class Solution {
public:
    int MoreThanHalfNum_Solution(vector<int> numbers) {
    	if (numbers.size() == 0) {
            return 0;
        }

        int target = numbers[0];
        unsigned int cnt = 1;

        for (int i = 1; i < numbers.size(); ++i) {
            if (target == numbers[i]) {
                cnt++;
            } else {
                cnt--;
            }

            if (cnt == 0) {
                target = numbers[i];
                cnt = 1;
            }
        }

        cnt = 0;

        for (int i = 0; i < numbers.size(); ++i) {
            if (target == numbers[i]) {
                cnt++;
            }
        }

        if (2 * cnt > numbers.size()) {
            return target;
        } else {
            return 0;
        }
    }
};

File: src/java/MoreThanHalfNum.java
import java.util.HashMap;

public class Solution {
    public int MoreThanHalfNum_Solution(int [] array) {
        HashMap<Integer, Integer> map = new HashMap<>();
        if (array == null || array.length == 0) return 0;

        for (int i = 0; i < array.length; i++) {
            Integer num = map.get(array[i]);
            if (num == null) {
                map.put(array[i], 1);
                num = 1;
            } else {
                map.put(array[i], ++num);
            }

            if (num > (array.length / 2)) {
                return array[i];
            }
        }

        return 0;
    }
}
